### PR TITLE
H5::Exception Handling

### PR DESCRIPTION
**Pull request by ghost**
### Description of work
When an H5::Exception is thrown by mantid code, usually for reasons relating to file-access permissions, the result is the cryptic message:  "unidentified C++ exception"  -- this is not useful!

Some cppchecks were broken due to file modifications and were fixed by either updating line numbers or handling the suppression at the root cause.

[3105: mantid catch handlers for H5::Exception](https://ornlrse.clm.ibmcloud.com/ccm/resource/itemName/com.ibm.team.workitem.WorkItem/3105)

#### Summary of work

`H5::Exception` is registered with `boost::python` and the caught exception error stack contents is forwarded.

### To test:

The following script should produce an improved error message:
```
import os
from mantid.simpleapi import *

target = os.path.expanduser('~/unwritable_directory')
os.mkdir(target, mode=0o500)

table = CreateEmptyTableWorkspace()
# table = mtd[tableWS]
table.addColumn(type='int', name='detid')
table.addColumn(type='double', name='difc')
table.addColumn(type='double', name='difa')
table.addColumn(type='double', name='tzero')
table.addColumn(type='double', name='tofMin')
table.addRow({'detid': 1, 'difc': 1000.0, 'difa': 0.0, 'tzero': 500.0, 'tofMin': 1000.0})
SaveDiffCal(CalibrationWorkspace=table, Filename=target + '/table.h5')
```

Previously the output is:
```
RuntimeError: SaveDiffCal-v1: unidentifiable C++ exception
File "<string>", line 17, in <module>
File "/home/kat/mambaforge/envs/SNAPRed/lib/python3.10/site-packages/mantid/simpleapi.py", line 1076, in __call__
raise RuntimeError(msg) from e
```

And now the output is:
```
H5 Exception in execution of algorithm SaveDiffCal:
H5File constructor: H5Fcreate failed
RuntimeError: SaveDiffCal-v1:   #0: unable to synchronously create file
  #1: unable to create file
  #2: file create failed
  #3: file create failed
  #4: unable to create file
  #5: unable to open file: name = '/home/ktz/unwritable_directory/table.h5', tent_flags = 15
  #6: open failed
  #7: unable to open file: name = '/home/ktz/unwritable_directory/table.h5', errno = 13, error message = 'Permission denied', flags = 15, o_flags = c2

  File "<string>", line 15, in <module>
  File "/home/ktz/trunk/mantid/repository/Framework/PythonInterface/mantid/simpleapi.py", line 1090, in __call__
    raise RuntimeError(msg) from e

```

---

### Reviewer

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

#### Code Review

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Do the release notes conform to the [release notes guide](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html)?

#### Functional Tests

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### Gatekeeper

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.

### CODE REVIEW

**Comment by sf1919** [comment]
Does this replace #37756 ? If so then the previous one will need closing.

File: Framework/API/src/Algorithm.cpp [hunk 1]
```diff
@@ -735,6 +736,22 @@ bool Algorithm::executeInternal() {
     throw;
   }
 
+  catch (H5::Exception &ex) {
+
+    std::string errmsg;
+    errmsg.append(ex.getCFuncName()).append(": ").append(ex.getCDetailMsg());
+
+    m_runningAsync = false;
```
Comment (peterfpeterson) [inline Framework/API/src/Algorithm.cpp:744]: From the other branches, it looks like this belongs as the first line in the error handling.

File: Framework/PythonInterface/mantid/api/src/Exports/Algorithm.cpp [hunk 2]
```diff
@@ -90,6 +91,26 @@ void translateCancel(const Algorithm::CancelException &exc) {
...
 }
 
+/**
+ * @brief Map an H5::Exception to a Python RuntimeError. Its error stack will be reported
+ *
+ * @param exc - The caught H5::Exception. Its error stack is reported
+ */
+void translateH5Exception(const H5::Exception &exc) {
+  auto handleFrame = [](unsigned int n, const H5E_error_t *err, void *ss_) -> herr_t {
+    std::stringstream &strstream = *((std::stringstream *)ss_);
```
Comment (peterfpeterson) [inline Framework/PythonInterface/mantid/api/src/Exports/Algorithm.cpp:101]: `strstream` is a dangerous variable name. There was a good long time when that was an alternate implementation of `stringstream`. Please change this name.

**Review by peterfpeterson — CHANGES_REQUESTED** [review]
As an addition to the detailed comments, please add release notes (I'd call this a framework feature). 

The PR description would benefit from a mention that you squashed a bunch of cppcheck issues. It is secondary thing, but explains some of the changes.

File: Framework/PythonInterface/mantid/api/src/Exports/Algorithm.cpp (hunk 2 above)
Comment (ktactac-ornl) [reply Framework/PythonInterface/mantid/api/src/Exports/Algorithm.cpp]: `ss_ -> ss_arg`
`strstream -> ss_`

File: Framework/API/src/Algorithm.cpp (hunk 1 above)
Comment (ktactac-ornl) [reply Framework/API/src/Algorithm.cpp:740]: Done

**Review by peterfpeterson — APPROVED** [review]

**Pull request merged**